File: src/dere_vault/scripts/concept_search.py
#!/usr/bin/env python3
"""Search permanent notes for similar concepts to prevent duplicates."""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

try:
    from .detect_vault import find_vault_root
except ImportError:
    from detect_vault import find_vault_root


def normalize_text(text: str) -> str:
    """Normalize text for comparison."""
    return text.lower().strip()


def extract_frontmatter_tags(content: str) -> list[str]:
    """Extract tags from YAML frontmatter."""
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return []

    frontmatter = match.group(1)
    tags = []
    in_tags = False

    for line in frontmatter.split("\n"):
        if line.startswith("tags:"):
            in_tags = True
            continue
        if in_tags:
            if line.startswith("  - "):
                tags.append(line.replace("  - ", "").strip())
            elif not line.startswith(" "):
                break

    return tags


def extract_title(content: str) -> str | None:
    """Extract title from frontmatter or first heading."""
    # Try frontmatter first
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if match:
        title_match = re.search(r"^title:\s*[\"']?(.+?)[\"']?\s*$", match.group(1), re.MULTILINE)
        if title_match:
            return title_match.group(1)

    # Fall back to first heading
    heading_match = re.search(r"^#\s+(.+)$", content, re.MULTILINE)
    if heading_match:
        return heading_match.group(1)

    return None


def is_permanent_note(content: str) -> bool:
    """Check if note is a permanent note."""
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return False

    frontmatter = match.group(1)
    return "type: permanent" in frontmatter


def calculate_similarity(query: str, note_title: str, note_tags: list[str], note_content: str) -> float:
    """Calculate similarity score between query and note (0.0-1.0)."""
    query_norm = normalize_text(query)
    title_norm = normalize_text(note_title)
    content_norm = normalize_text(note_content)

    score = 0.0

    # Exact title match
    if query_norm == title_norm:
        score += 1.0
    # Title contains query
    elif query_norm in title_norm:
        score += 0.7
    # Query contains title words
    elif any(word in query_norm for word in title_norm.split() if len(word) > 3):
        score += 0.5

    # Tag overlap
    query_words = set(query_norm.split())
    tag_words = set(" ".join(note_tags).lower().split())
    if query_words & tag_words:
        score += 0.3

    # Content contains query
    if query_norm in content_norm:
        score += 0.2

    return min(score, 1.0)


def search_concepts(vault_path: Path, query: str, threshold: float = 0.3) -> list[dict]:
    """Search for similar concepts in permanent notes."""
    results = []

    # Search all markdown files
    for note_path in vault_path.rglob("*.md"):
        try:
            content = note_path.read_text(encoding="utf-8")
        except Exception:
            continue

        # Only process permanent notes
        if not is_permanent_note(content):
            continue

        title = extract_title(content)
        if not title:
            continue

        tags = extract_frontmatter_tags(content)

        # Calculate similarity
        similarity = calculate_similarity(query, title, tags, content)

        if similarity >= threshold:
            results.append({
                "path": note_path.relative_to(vault_path),
                "title": title,
                "tags": tags,
                "similarity": similarity,
            })

    # Sort by similarity (descending)
    results.sort(key=lambda x: x["similarity"], reverse=True)

    return results


def main() -> int:
    """CLI entry point."""
    parser = argparse.ArgumentParser(description="Search permanent notes for similar concepts")
    parser.add_argument("query", help="Concept to search for")
    parser.add_argument("--threshold", type=float, default=0.3, help="Minimum similarity score (0.0-1.0)")
    parser.add_argument("--vault-path", type=Path, help="Path to vault (auto-detected if not provided)")
    parser.add_argument("--limit", type=int, default=10, help="Maximum number of results")

    args = parser.parse_args()

    # Find vault
    vault_path = args.vault_path
    if not vault_path:
        vault_path = find_vault_root()
        if not vault_path:
            print("Error: Could not find vault. Specify --vault-path", file=sys.stderr)
            return 1

    # Search
    results = search_concepts(vault_path, args.query, args.threshold)

    if not results:
        print(f"No similar concepts found for: {args.query}")
        return 0

    # Display results
    print(f"Found {len(results)} similar concept(s) for: {args.query}\n")

    for i, result in enumerate(results[:args.limit], 1):
        print(f"{i}. {result['title']} (similarity: {result['similarity']:.2f})")
        print(f"   Path: {result['path']}")
        if result['tags']:
            print(f"   Tags: {', '.join(result['tags'])}")
        print()

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/dere_plugins/dere_vault/scripts/concept_search.ts
import { readFile } from "node:fs/promises";
import { relative } from "node:path";

import { findVaultRoot } from "./detect_vault.js";

function normalizeText(text: string): string {
  return text.toLowerCase().trim();
}

function extractFrontmatterTags(content: string): string[] {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    return [];
  }
  const frontmatter = match[1];
  const tags: string[] = [];
  let inTags = false;

  for (const line of frontmatter.split("\n")) {
    if (line.startsWith("tags:")) {
      inTags = true;
      continue;
    }
    if (inTags) {
      if (line.startsWith("  - ")) {
        tags.push(line.replace("  - ", "").trim());
      } else if (!line.startsWith(" ")) {
        break;
      }
    }
  }

  return tags;
}

function extractTitle(content: string): string | null {
  const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---/);
  if (frontmatterMatch) {
    const titleMatch = frontmatterMatch[1].match(/^title:\s*["']?(.+?)["']?\s*$/m);
    if (titleMatch?.[1]) {
      return titleMatch[1];
    }
  }

  const headingMatch = content.match(/^#\s+(.+)$/m);
  if (headingMatch?.[1]) {
    return headingMatch[1];
  }

  return null;
}

function isPermanentNote(content: string): boolean {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    return false;
  }
  return match[1].includes("type: permanent");
}

function calculateSimilarity(
  query: string,
  noteTitle: string,
  noteTags: string[],
  noteContent: string,
): number {
  const queryNorm = normalizeText(query);
  const titleNorm = normalizeText(noteTitle);
  const contentNorm = normalizeText(noteContent);

  let score = 0;

  if (queryNorm === titleNorm) {
    score += 1;
  } else if (titleNorm.includes(queryNorm)) {
    score += 0.7;
  } else if (titleNorm.split(/\s+/).some((word) => word.length > 3 && queryNorm.includes(word))) {
    score += 0.5;
  }

  const queryWords = new Set(queryNorm.split(/\s+/));
  const tagWords = new Set(noteTags.join(" ").toLowerCase().split(/\s+/));
  for (const word of queryWords) {
    if (tagWords.has(word)) {
      score += 0.3;
      break;
    }
  }

  if (contentNorm.includes(queryNorm)) {
    score += 0.2;
  }

  return Math.min(score, 1);
}

async function* walk(dir: string): AsyncGenerator<string> {
  const { readdir } = await import("node:fs/promises");
  let items;
  try {
    items = await readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const item of items) {
    const fullPath = join(dir, item.name);
    if (item.isDirectory()) {
      yield* walk(fullPath);
    } else if (item.isFile() && item.name.endsWith(".md")) {
      yield fullPath;
    }
  }
}

type ConceptResult = {
  path: string;
  title: string;
  tags: string[];
  similarity: number;
};

async function searchConcepts(
  vaultPath: string,
  query: string,
  threshold: number,
): Promise<ConceptResult[]> {
  const results: ConceptResult[] = [];

  for await (const notePath of walk(vaultPath)) {
    let content: string;
    try {
      content = await readFile(notePath, "utf-8");
    } catch {
      continue;
    }

    if (!isPermanentNote(content)) {
      continue;
    }

    const title = extractTitle(content);
    if (!title) {
      continue;
    }

    const tags = extractFrontmatterTags(content);
    const similarity = calculateSimilarity(query, title, tags, content);

    if (similarity >= threshold) {
      results.push({
        path: relative(vaultPath, notePath),
        title,
        tags,
        similarity,
      });
    }
  }

  results.sort((a, b) => b.similarity - a.similarity);
  return results;
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.error(
      "Usage: concept_search.ts <query> [--threshold 0.3] [--vault-path path] [--limit 10]",
    );
    process.exit(1);
  }

  const query = args[0];
  let threshold = 0.3;
  let vaultPath: string | null = null;
  let limit = 10;

  for (let i = 1; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === "--threshold") {
      threshold = Number.parseFloat(args[i + 1] ?? "0.3");
      i += 1;
    } else if (arg === "--vault-path") {
      vaultPath = args[i + 1] ?? null;
      i += 1;
    } else if (arg === "--limit") {
      limit = Number.parseInt(args[i + 1] ?? "10", 10);
      i += 1;
    }
  }

  const resolvedVaultPath = vaultPath ?? findVaultRoot();
  if (!resolvedVaultPath) {
    console.error("Error: Could not find vault. Specify --vault-path");
    process.exit(1);
  }

  const results = await searchConcepts(resolvedVaultPath, query, threshold);
  if (results.length === 0) {
    console.log(`No similar concepts found for: ${query}`);
    process.exit(0);
  }

  console.log(`Found ${results.length} similar concept(s) for: ${query}\n`);
  const slice = results.slice(0, limit);
  slice.forEach((result, index) => {
    console.log(`${index + 1}. ${result.title} (similarity: ${result.similarity.toFixed(2)})`);
    console.log(`   Path: ${result.path}`);
    if (result.tags.length) {
      console.log(`   Tags: ${result.tags.join(", ")}`);
    }
    console.log("");
  });
}

if (import.meta.main) {
  void main();
}
